lib/utilities/request.js

```diff
@@ -0,0 +1,137 @@
+    if (!endpoint || typeof endpoint === 'function') {
+        throw new Error('Missing target endpoint');
+    }
+
+    let cb;
+    let opts = {};
+    if (typeof options === 'function') {
+        cb = jsutil.once(options);
+    } else if (typeof options === 'object') {
+        opts = options;
+        if (typeof callback === 'function') {
+            cb = jsutil.once(callback);
+        }
+    }
+
+    if (typeof cb !== 'function') {
+        throw new Error('Missing request callback');
+    }
+
+    if (!(endpoint instanceof url.URL || typeof endpoint === 'string')) {
+        return cb(new Error(`Invalid URI ${endpoint}`));
+    }
+
+    if (!opts.method) {
+        opts.method = 'GET';
+    } else if (!validVerbs.includes(opts.method)) {
+        return cb(new Error(`Invalid Method ${opts.method}`));
+    }
+
+    let reqParams = endpoint;
+    if (typeof reqParams === 'string') {
+        try {
+            reqParams = url.parse(endpoint);
+        } catch (error) {
+            return cb(error);
+        }
+    }
+    reqParams.method = opts.method;
+
+    let request;
+    if (reqParams.protocol === 'http:') {
+        request = http;
+        const httpProxy = process.env.HTTP_PROXY || process.env.http_proxy;
+        if (!proxyCompareUrl(reqParams.hostname) && httpProxy) {
+            reqParams.agent = new HttpProxyAgent(url.parse(httpProxy));
+        }
+    } else if (reqParams.protocol === 'https:') {
+        request = https;
+        const httpsProxy = process.env.HTTPS_PROXY || process.env.https_proxy;
+        if (!proxyCompareUrl(reqParams.hostname) && httpsProxy) {
+            reqParams.agent = new HttpsProxyAgent(url.parse(httpsProxy));
+        }
+    } else {
+        return cb(new Error(`Invalid Protocol ${reqParams.protocol}`));
+    }
+
+    reqParams.headers = Object.assign({}, opts.headers);
+    let data;
+    if (opts.body) {
+        if (typeof opts.body === 'object') {
+            data = JSON.stringify(opts.body);
+            if (!reqParams.headers['content-type'] &&
+                !reqParams.headers['Content-Type']) {
+                reqParams.headers['content-type'] = 'application/json';
+            }
+        } else {
+            data = opts.body;
+        }
+        reqParams.headers['Content-Length'] = Buffer.byteLength(data);
+    }
+
+    const req = request.request(reqParams);
+    req.on('error', cb);
+    req.on('response', res => {
+        let rawData  = '';
+        res.on('data', chunk => { rawData += chunk; });
+        res.on('end', () => {
+            if (opts.json && rawData) {
+                try {
+                    const parsed = JSON.parse(rawData);
+                    return cb(null, res, parsed);
+                } catch (err) {
+                    // invalid json response
+                    return cb(err, res, null);
+                }
+            }
+            return cb(null, res, rawData);
+        });
+    });
+    if (updateVerbs.includes(opts.method)) {
```

Comment:
I'd also check whether `data !== undefined` since one might want to do a POST/PUT without a body.